contracts/marketplace/lib.rs

```diff
@@ -0,0 +1,299 @@
+            Self::emit_event(self.env(), Event::Reset(Reset {}));
+
+            Ok(())
+        }
+
+        fn current_price(&self) -> Balance {
+            let block = self.env().block_number();
+            let elapsed = block.saturating_sub(self.current_start_block.into());
+            self.average_price()
+                .saturating_mul(self.sale_multiplier)
+                .saturating_sub(self.per_block_reduction().saturating_mul(elapsed.into()))
+                .max(self.min_price)
+        }
+
+        fn per_block_reduction(&self) -> Balance {
+            self.average_price()
+                .saturating_div(self.auction_length.into())
+                .max(1u128)
+        }
+
+        fn take_payment(&self, from: AccountId, amount: Balance) -> Result<(), Error> {
+            build_call::<Environment>()
+                .call_type(Call::new().callee(self.reward_token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(TRANSFER_FROM_GAME_TOKEN_SELECTOR))
+                        .push_arg(from)
+                        .push_arg(self.this())
+                        .push_arg(amount)
+                        .push_arg(DUMMY_DATA),
+                )
+                .call_flags(CallFlags::default().set_allow_reentry(true))
+                .returns::<Result<(), PSP22Error>>()
+                .fire()??;
+
+            Ok(())
+        }
+
+        fn give_ticket(&self, to: AccountId) -> Result<(), Error> {
+            build_call::<Environment>()
+                .call_type(Call::new().callee(self.ticket_token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(TRANSFER_TICKET_SELECTOR))
+                        .push_arg(to)
+                        .push_arg(1u128)
+                        .push_arg(DUMMY_DATA),
+                )
+                .returns::<Result<(), PSP22Error>>()
+                .fire()??;
+
+            Ok(())
+        }
+
+        fn ticket_balance(&self) -> Result<Balance, Error> {
+            let balance = build_call::<Environment>()
+                .call_type(Call::new().callee(self.ticket_token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(TICKET_BALANCE_SELECTOR))
+                        .push_arg(self.this()),
+                )
+                .returns::<Balance>()
+                .fire()?;
+
+            Ok(balance)
+        }
+
+        fn ensure_role(role: Role) -> Result<(), Error> {
+            <Self as AccessControlled>::check_role(
+                AccountId::from(ACCESS_CONTROL_PUBKEY),
+                Self::env().caller(),
+                role,
+                |reason| reason.into(),
+                |role| Error::missing_role(role),
+            )
+        }
+
+        fn initializer() -> Role {
+            let code_hash = Self::env()
+                .own_code_hash()
+                .expect("Failure to retrieve code hash.");
+            Initializer(code_hash)
```

Comment:
Here we're calling the `Role` enum's variant directly, I'd prefer to be consistent and call it like the rest (like `admin` below)  - `Role::Initializer`.